src/uct/tcp/tcp_ep.c

```diff
@@ -591,21 +624,41 @@ static inline ssize_t uct_tcp_ep_sendv(uct_tcp_ep_t *ep)
     return sent_length;
 }
 
-void uct_tcp_ep_dropped_connect_print_error(uct_tcp_ep_t *ep, int io_errno)
+ucs_status_t uct_tcp_ep_handle_dropped_connect_error(uct_tcp_ep_t *ep, int io_errno)
 {
-    /* if connection establishment failes, the system limits
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    ucs_status_t status;
+
+    /* if connection establishment fails, the system limits
      * may not be big enough */
     if (((ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
          (ep->conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
          (ep->conn_state == UCT_TCP_EP_CONN_STATE_WAITING_REQ)) &&
         ((io_errno == ECONNRESET) || (io_errno == ECONNREFUSED) ||
          /* connection establishment procedure timed out */
          (io_errno == ETIMEDOUT))) {
-        ucs_error("try to increase \"net.core.somaxconn\", "
-                  "\"net.core.netdev_max_backlog\", "
-                  "\"net.ipv4.tcp_max_syn_backlog\" to the maximum value "
-                  "on the remote node");
+        uct_tcp_ep_mod_events(ep, 0, ep->events);
+        uct_tcp_ep_close_fd(&ep->fd);
+
+        uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CLOSED);
+
+        status = uct_tcp_ep_create_socket_and_connect(iface, NULL, &ep);
+        if (status != UCS_OK) {
+            ucs_error("try to increase \"net.core.somaxconn\", "
+                      "\"net.core.netdev_max_backlog\", "
+                      "\"net.ipv4.tcp_max_syn_backlog\" to the maximum value "
+                      "on the remote node or increase %s%s%s=%u to do more "
+                      "more connection establishment attempts",
```

Comment:
can remove this line - error message is already too long

Reply:
done